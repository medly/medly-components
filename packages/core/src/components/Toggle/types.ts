import { FontVariants, FontWeights, ToggleSizes } from '@medly-components/theme';
import { HTMLProps, Omit } from '@medly-components/utils';
import { ReactElement } from 'react';
import { LabelPositions } from '../Label/types';

type DivProps = Omit<HTMLProps<HTMLInputElement>, 'size'>;
type InputProps = Omit<HTMLProps<HTMLInputElement>, 'size' | 'type'>;

export interface ToggleWrapperProps extends DivProps {
    size?: ToggleSizes;
}

export interface ToggleProps extends InputProps {
    /** Toggle label */
    label?: string;
    /** Toggle size */
    size?: ToggleSizes;
    /** Takes full width of the parent component */
    fullWidth?: boolean;
    /** Label position */
    labelPosition?: LabelPositions;
    /** Label variant */
    labelVariant?: FontVariants;
    /** Label color */
    labelColor?: string;
    /** Label weight */
    labelWeight?: FontWeights;
    /** Use if required in any form for example user needs to check 'I Agree' */
    required?: boolean;
    /** Icon to show when toggle is in OFF state */
    iconRight?: ReactElement;
    /** Icon to show when toggle is in ON state */
    iconLeft?: ReactElement;
}
