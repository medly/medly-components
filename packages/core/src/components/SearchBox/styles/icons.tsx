import { SvgIcon } from '@medly-components/icons';
import { css, styled } from '@medly-components/utils';
import { SearchBoxProps } from '../types';
import { getIconWrapperStyle } from './utils';

export const CloseIconWrapper = styled.span<SearchBoxProps & { showCloseIcon?: boolean; hasCustomSearchFilter?: boolean }>`
    ${getIconWrapperStyle};
    border-right: ${({ theme, showCloseIcon }) => showCloseIcon && `0.1rem solid ${theme.colors.grey[200]}`};
    margin: 0.6rem 0;
    padding-right: ${({ hasCustomSearchFilter }) => hasCustomSearchFilter && '0.4rem'};
    ${SvgIcon} {
        padding: ${({ theme, size }) => theme.searchBox.closeIcon.padding[size]};
        border-radius: ${({ theme }) => theme.searchBox.closeIcon.borderRadius};
        * {
            fill: ${({ theme }) => theme.searchBox.closeIcon.color.default};
        }
        &:hover {
            background: ${({ theme }) => theme.searchBox.closeIcon.bgColor.hovered};
            * {
                fill: ${({ theme }) => theme.searchBox.closeIcon.color.hovered};
            }
        }
        &:active {
            background: ${({ theme }) => theme.searchBox.closeIcon.bgColor.pressed};
            * {
                fill: ${({ theme }) => theme.searchBox.closeIcon.color.pressed};
            }
        }
    }
`;

export const SearchIconWrapper = styled.span<SearchBoxProps & { areOptionsVisible?: boolean; isTyping?: boolean }>`
    ${getIconWrapperStyle};
    ${SvgIcon} {
        padding: ${({ theme, size }) => theme.searchBox.searchIcon.padding[size]};
        border-radius: ${({ theme }) => theme.searchBox.searchIcon.borderRadius};
        * {
            fill: ${({ theme: { searchBox }, areOptionsVisible }) => searchBox.searchIcon.color[areOptionsVisible ? 'active' : 'default']};
        }
        &:hover {
            background: ${({ theme: { searchBox }}) => searchBox.searchIcon.bgColor.hovered};
        }
        &:active {
            background: ${({ theme }) => theme.searchBox.searchIcon.bgColor.pressed};
            * {
                fill: ${({ theme }) => theme.searchBox.searchIcon.color.pressed};
            }
        }
    }
`;

const customSearchActiveStyle = css`
    ${SvgIcon} {
        background: ${({ theme }) => theme.searchBox.expandIcon.bgColor.active};
        path {
            transform: rotate(180deg);
        }

        * {
            fill: ${({ theme }) => theme.searchBox.expandIcon.color.active};
        }

        &:hover {
            background: ${({ theme }) => theme.searchBox.expandIcon.bgColor.active};
        }
    }
`

export const ExpandIconWrapper = styled.span<SearchBoxProps & { isCustomSearchActive?: boolean }>`
    ${getIconWrapperStyle}
    ${SvgIcon} {
        padding: ${({ theme, size }) => theme.searchBox.expandIcon.padding[size]};
        border-radius: ${({ theme }) => theme.searchBox.expandIcon.borderRadius};
        path {
            transition: transform 0.2s ease-out;
            transform-origin: center;
        }
        * {
            fill: ${({ theme }) => theme.searchBox.expandIcon.color.default};
        }
        &:hover {
            background: ${({ theme }) => theme.searchBox.expandIcon.bgColor.hovered};
        }
        &:active {
            background: ${({ theme }) => theme.searchBox.expandIcon.bgColor.pressed};
            * {
                fill: ${({ theme }) => theme.searchBox.expandIcon.color.pressed};
            }
        }
    }
    ${({ isCustomSearchActive }) => isCustomSearchActive && customSearchActiveStyle};
`;
