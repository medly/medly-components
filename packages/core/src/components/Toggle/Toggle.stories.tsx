import { defaultTheme, ToggleTheme } from '@medly-components/theme';
import React from 'react';
import { Props } from './types';

export const labelPositions: Props['labelPosition'][] = ['left', 'right', 'top', 'bottom'];
export const sizes: Props['size'][] = ['XS', 'S', 'M', 'L', 'XL'];

export const ThemeInterface: React.FC<ToggleTheme> = () => null;
ThemeInterface.defaultProps = {
    ...defaultTheme.toggle
};


