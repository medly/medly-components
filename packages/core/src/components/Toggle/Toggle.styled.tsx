import { defaultTheme } from '@medly-components/theme';
import { WithThemeProp } from '@medly-components/utils';
import styled, { css } from 'styled-components';
import FieldWithLabel from '../FieldWithLabel';
import { ToggleProps, ToggleWrapperProps } from './types';

const getHeight = ({ theme, size }: ToggleProps & WithThemeProp) =>
    size ? theme.toggle.sizes[size] : theme.toggle.sizes[theme?.toggle.defaultSize];

const getWidth = (props: ToggleProps & WithThemeProp) => `calc(${getHeight(props)} * 2)`;

const IconStyle = css`
    color: ${({ theme }) => theme.toggle.iconColor};
    transition: 0.2s ease-in-out;
    position: absolute;
    svg {
        height: ${props => `calc(${getHeight(props)} - 1rem) `};
        path {
            fill: ${({ theme }) => theme.toggle.iconColor};
        }
    }
`;

export const Wrapper = styled(FieldWithLabel.Field)<ToggleWrapperProps>`
    position: relative;
    width: ${getWidth};
    height: ${getHeight};
    border-radius: 1.5rem;
    background-color: ${({ theme }) => theme.toggle.bgColor};

    &:focus-within {
        box-shadow: 0 0 0 1pt ${({ theme }) => theme.toggle.outlineColor};
    }

    & > * {
        width: 100%;
        height: 100%;
        border-radius: 1.5rem;
    }
`;

Wrapper.defaultProps = {
    theme: defaultTheme
};

export const Switch = styled('div')<ToggleWrapperProps>`
    position: relative;
    width: 100%;
    overflow: hidden;
    display: flex;
    align-items: flex-end;
`;

export const OffIcon = styled('div')<ToggleWrapperProps>`
    ${IconStyle}
    right: 0;
`;

export const OnIcon = styled('div')<ToggleWrapperProps>`
    ${IconStyle}
    opacity: 0;
    left: ${props => `calc((-${getHeight(props)}) + 0.3rem) `};
`;

export const Circle = styled('div')<ToggleWrapperProps>`
    z-index: 1;
    position: absolute;
    height: ${props => `calc(${getHeight(props)} - 0.6rem) `};
    width: ${props => `calc(${getHeight(props)} - 0.6rem) `};
    margin: 0.3rem;
    border-radius: 50%;
    background: ${({ theme }) => theme.colors.white};
    box-shadow: 0.1rem 0.3rem 0.3rem 0.1rem rgba(0, 0, 0, 0.2);
    transition: 0.2s;
`;

Circle.defaultProps = {
    theme: defaultTheme
};

export const Checkbox = styled('input').attrs({ type: 'checkbox' })<ToggleProps>`
    opacity: 0;
    z-index: 2;
    margin: 0;
    position: absolute;
    cursor: pointer;

    &:focus {
        outline: none;
    }

    &:disabled {
        cursor: not-allowed;
        & + ${Switch} {
            background: ${({ theme }) => theme.toggle.disabledBgColor};
        }
    }

    &:checked + ${Switch} {
        background: ${({ theme }) => theme.toggle.checkedBgColor};

        ${Circle} {
            margin-left: ${props => `calc((${getHeight(props)}) + 0.3rem) `};
        }

        ${OffIcon} {
            color: ${({ theme }) => theme.toggle.iconColor};
            opacity: 0;
            right: ${props => `calc((-${getHeight(props)}) + 0.3rem) `};
        }

        ${OnIcon} {
            opacity: 1;
            position: absolute;
            left: 0;
        }
    }
`;
