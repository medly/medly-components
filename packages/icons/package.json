{
    "name": "@medly-components/icons",
    "version": "1.7.17",
    "sideEffects": false,
    "main": "dist/cjs/index.js",
    "module": "dist/es/index.js",
    "types": "dist/es/index.d.ts",
    "files": [
        "dist"
    ],
    "publishConfig": {
        "access": "public"
    },
    "license": "MIT",
    "scripts": {
        "create-icons": "ts-node ./src/createIcons.ts",
        "precreate-icons": "rimraf ./src/icons && rimraf ./src/index.ts && rimraf ./src/icons.stories.mdx"
    },
    "peerDependencies": {
        "react": "16.x || 17.x",
        "react-dom": "16.x || 17.x",
        "styled-components": "^5.1.0"
    },
    "dependencies": {
        "@medly-components/theme": "2.18.0",
        "@medly-components/utils": "1.14.8"
    }
}
