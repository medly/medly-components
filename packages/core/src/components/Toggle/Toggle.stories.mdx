import { Toggle } from './Toggle';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import * as stories from './Toggle.stories';
import { boolean, select } from '@storybook/addon-knobs';
import { sizes, labelPositions, ThemeInterface } from './Toggle.stories';
import { LockOpenIcon, LockOutlineIcon } from '@medly-components/icons';

<Meta title="Core" component={Toggle} parameters={{ jest: ['Toggle.test.tsx'] }} />

# Toggle

The `Toggle` component is a control that you must use to switch quickly between two possible states. It is commonly used for “on/off” switches.
You can use it as an alternative for the `Checkbox` component. It semantically uses the inputs checkbox variant.

<Preview withToolbar>
    <Story name="Toggle">
        <Toggle
            size={select('Size', sizes, 'S')}
            label="Switch mode"
            fullWidth={boolean('Full Width', false)}
            disabled={boolean('Disabled', false)}
            labelPosition={select('Label Position', labelPositions, 'right')}
        />
    </Story>
</Preview>

## With Icon

You can add icon to the toggle component by passing the `iconLeft` and `iconRight` prop

<Preview withToolbar>
    <Story name="Toggle with icon">
        <Toggle
            size={select('Size', sizes, 'S')}
            label="Switch mode"
            fullWidth={boolean('Full Width', false)}
            disabled={boolean('Disabled', false)}
            labelPosition={select('Label Position', labelPositions, 'right')}
            iconLeft={<LockOutlineIcon />}
            iconRight={<LockOpenIcon />}
        />
    </Story>
</Preview>

## Props

This component offers the following props mentioned in the table below:

<Props of={Toggle} />

## ToggleTheme

This component offers the following themes mentioned in the table below:

<Props of={stories.ThemeInterface} />
