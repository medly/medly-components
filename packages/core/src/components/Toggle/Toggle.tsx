import { WithStyle } from '@medly-components/utils';
import { cloneElement, FC, forwardRef, memo, useCallback, useMemo } from 'react';
import FieldWithLabel from '../FieldWithLabel';
import * as Styled from './Toggle.styled';
import { ToggleProps } from './types';

const Component: FC<ToggleProps> = memo(
    forwardRef((props, ref) => {
        const {
            id,
            size,
            label,
            required,
            labelPosition,
            labelVariant,
            labelWeight,
            labelColor,
            fullWidth,
            onChange,
            iconLeft,
            iconRight,
            ...restProps
        } = props;

        const changeHandler = useCallback(
                (e: any) => {
                    e.stopPropagation();
                    onChange && onChange(e);
                },
                [onChange]
            ),
            toggleId = useMemo(() => id || label?.toLocaleLowerCase().replace(/ +/g, '-') || 'medly-toggle-checkbox', [id, label]);

        return (
            <FieldWithLabel id={toggleId} fieldWithMaxContent {...{ fullWidth, labelPosition }}>
                {label && (
                    <FieldWithLabel.Label
                        showPointer={!restProps.disabled}
                        {...{ required, labelPosition, labelVariant, labelWeight, labelColor }}
                        htmlFor={toggleId}
                    >
                        {label}
                    </FieldWithLabel.Label>
                )}
                <Styled.Wrapper size={size} disabled={restProps.disabled}>
                    {/* @ts-ignore */}
                    <Styled.Checkbox ref={ref} id={label} size={size} required={required} onChange={changeHandler} {...restProps} />
                    <Styled.Switch>
                        {iconLeft && <Styled.OnIcon size={size}>{cloneElement(iconLeft, { size })}</Styled.OnIcon>}
                        <Styled.Circle size={size} />
                        {iconRight && <Styled.OffIcon size={size}>{cloneElement(iconRight, { size })}</Styled.OffIcon>}
                    </Styled.Switch>
                </Styled.Wrapper>
            </FieldWithLabel>
        );
    })
);

Component.displayName = 'Toggle';
Component.defaultProps = {
    label: '',
    required: false,
    labelPosition: 'right'
};
export const Toggle: FC<ToggleProps> & WithStyle = Object.assign(Component, { Style: FieldWithLabel.Style });
