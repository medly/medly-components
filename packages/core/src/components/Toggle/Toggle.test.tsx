import { LockOpenIcon, LockOutlineIcon } from '@medly-components/icons';
import { fireEvent, render, screen } from '@test-utils';
import { Toggle } from './Toggle';

describe('Checkbox component', () => {
    it('should render correctly with all the default props', () => {
        const { container } = render(<Toggle name="gender" />);
        expect(container).toMatchSnapshot();
    });

    it('should render correctly with all the props given', () => {
        const { container } = render(
            <Toggle disabled fullWidth name="gender" label="Female" size="M" labelPosition="right" id="gender-toggle" />
        );
        expect(container).toMatchSnapshot();
    });

    it('should call onChange handler on click on checkbox', () => {
        const mockOnChange = jest.fn();
        render(<Toggle onChange={mockOnChange} disabled name="gender" label="Female" size="M" labelPosition="right" />);
        fireEvent.click(screen.getByRole('checkbox'));
        expect(mockOnChange).toBeCalled();
    });

    it('should show the toggle button icons when icons are passed', () => {
        const mockOnChange = jest.fn();
        render(
            <Toggle
                onChange={mockOnChange}
                disabled
                name="gender"
                label="Female"
                size="M"
                labelPosition="right"
                iconLeft={<LockOpenIcon data-testid="lock-open-icon" />}
                iconRight={<LockOutlineIcon data-testid="lock-close-icon" />}
            />
        );
        expect(screen.getByTestId('lock-open-icon')).toBeInTheDocument();
        expect(screen.getByTestId('lock-close-icon')).toBeInTheDocument();
    });
});
