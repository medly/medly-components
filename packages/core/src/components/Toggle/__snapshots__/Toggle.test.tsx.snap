// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component should render correctly with all the default props 1`] = `
.c0 {
  margin: 0.8rem 0.8rem 0.8rem 0;
  display: inline-grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  grid-template-columns: max-content minmax(max-content,25%);
  grid-template-areas: 'field label' 'description random';
  grid-template-columns: auto;
  grid-template-areas: 'field' 'description';
}

.c0 .sc-12atfcu-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c1 {
  grid-area: field;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-left: 0px;
  position: relative;
  width: calc(24px * 2);
  height: 24px;
  border-radius: 1.5rem;
  background-color: #dfe4e9;
}

.c1:focus-within {
  box-shadow: 0 0 0 1pt #012040;
}

.c1 > * {
  width: 100%;
  height: 100%;
  border-radius: 1.5rem;
}

.c4 {
  position: relative;
  width: 100%;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c6 {
  z-index: 1;
  position: absolute;
  height: calc(24px - 0.6rem);
  width: calc(24px - 0.6rem);
  margin: 0.3rem;
  border-radius: 50%;
  background: #ffffff;
  box-shadow: 0.1rem 0.3rem 0.3rem 0.1rem rgba(0,0,0,0.2);
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c2 {
  opacity: 0;
  z-index: 2;
  margin: 0;
  position: absolute;
  cursor: pointer;
}

.c2:focus {
  outline: none;
}

.c2:disabled {
  cursor: not-allowed;
}

.c2:disabled + .c3 {
  background: #C7D0D8;
}

.c2:checked + .c3 {
  background: #012040;
}

.c2:checked + .c3 .c5 {
  margin-left: calc((24px) + 0.3rem);
}

.c2:checked + .c3 .sc-1oxfop0-2 {
  color: #ffffff;
  opacity: 0;
  right: calc((-24px) + 0.3rem);
}

.c2:checked + .c3 .sc-1oxfop0-3 {
  opacity: 1;
  position: absolute;
  left: 0;
}

<div>
  <div
    class="c0"
    id="medly-toggle-checkbox"
  >
    
    <div
      class="sc-12atfcu-0 c1"
    >
      <input
        class="c2"
        id=""
        name="gender"
        type="checkbox"
      />
      <div
        class="c3 c4"
      >
        <div
          class="c5 c6"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Checkbox component should render correctly with all the props given 1`] = `
.c1 {
  color: #13181D;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  margin: 0px 0px 0px 0.8rem;
  cursor: not-allowed;
}

.c2 {
  grid-area: label;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 {
  margin: 0.8rem 0;
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  grid-template-columns: max-content minmax(max-content,25%);
  grid-template-areas: 'field label' 'description random';
}

.c0 .sc-12atfcu-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c3 {
  grid-area: field;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-left: 0px;
  position: relative;
  width: calc(28px * 2);
  height: 28px;
  border-radius: 1.5rem;
  background-color: #dfe4e9;
}

.c3:focus-within {
  box-shadow: 0 0 0 1pt #012040;
}

.c3 > * {
  width: 100%;
  height: 100%;
  border-radius: 1.5rem;
}

.c6 {
  position: relative;
  width: 100%;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c8 {
  z-index: 1;
  position: absolute;
  height: calc(28px - 0.6rem);
  width: calc(28px - 0.6rem);
  margin: 0.3rem;
  border-radius: 50%;
  background: #ffffff;
  box-shadow: 0.1rem 0.3rem 0.3rem 0.1rem rgba(0,0,0,0.2);
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c4 {
  opacity: 0;
  z-index: 2;
  margin: 0;
  position: absolute;
  cursor: pointer;
}

.c4:focus {
  outline: none;
}

.c4:disabled {
  cursor: not-allowed;
}

.c4:disabled + .c5 {
  background: #C7D0D8;
}

.c4:checked + .c5 {
  background: #012040;
}

.c4:checked + .c5 .c7 {
  margin-left: calc((28px) + 0.3rem);
}

.c4:checked + .c5 .sc-1oxfop0-2 {
  color: #ffffff;
  opacity: 0;
  right: calc((-28px) + 0.3rem);
}

.c4:checked + .c5 .sc-1oxfop0-3 {
  opacity: 1;
  position: absolute;
  left: 0;
}

<div>
  <div
    class="c0"
    id="gender-toggle"
  >
    <label
      class="c1 c2"
      for="gender-toggle"
    >
      Female
    </label>
    <div
      class="sc-12atfcu-0 c3"
      disabled=""
    >
      <input
        class="c4"
        disabled=""
        id="Female"
        name="gender"
        type="checkbox"
      />
      <div
        class="c5 c6"
      >
        <div
          class="c7 c8"
        />
      </div>
    </div>
  </div>
</div>
`;
